pkg/image/importer/importer.go

```diff
@@ -0,0 +1,768 @@
+					Name:  defaultRef.Tag,
+					Image: cache[tag],
+				})
+			}
+		}
+	}
+
+	// for each repository we found, import all tags and digests
+	for key, repo := range repositories {
+		importRepositoryFromDocker(ctx, repo, limiter)
+		for _, tag := range repo.Tags {
+			j := manifestKey{repositoryKey: key}
+			j.value = tag.Name
+			if tag.Image != nil {
+				cache[j] = tag.Image
+			}
+			for _, index := range tags[j] {
+				if tag.Err != nil {
+					setImageImportStatus(isi, index, tag.Err)
+					continue
+				}
+				copied := *tag.Image
+				image := &isi.Status.Images[index]
+				ref := repo.Ref
+				ref.Tag, ref.ID = tag.Name, copied.Name
+				copied.DockerImageReference = ref.MostSpecific().Exact()
+				image.Tag = tag.Name
+				image.Image = &copied
+				image.Status.Status = unversioned.StatusSuccess
+			}
+		}
+		for _, digest := range repo.Digests {
+			j := manifestKey{repositoryKey: key}
+			j.value = digest.Name
+			if digest.Image != nil {
+				cache[j] = digest.Image
+			}
+			for _, index := range ids[j] {
+				if digest.Err != nil {
+					setImageImportStatus(isi, index, digest.Err)
+					continue
+				}
+				image := &isi.Status.Images[index]
+				copied := *digest.Image
+				ref := repo.Ref
+				ref.Tag, ref.ID = "", copied.Name
+				copied.DockerImageReference = ref.MostSpecific().Exact()
+				image.Image = &copied
+				image.Status.Status = unversioned.StatusSuccess
+			}
+		}
+	}
+}
+
+// importFromRepository imports the repository named on the ImageStreamImport, if any, importing up to maximumTags, and reporting
+// status on each image that is attempted to be imported. If the repository cannot be found or tags cannot be retrieved, the repository
+// status field is set.
+func importFromRepository(ctx *AuthContext, isi *api.ImageStreamImport, maximumTags int, cache map[manifestKey]*api.Image, limiter util.RateLimiter) {
+	if isi.Spec.Repository == nil {
+		return
+	}
+	isi.Status.Repository = &api.RepositoryImportStatus{}
+	status := isi.Status.Repository
+
+	spec := isi.Spec.Repository
+	from := spec.From
+	if from.Kind != "DockerImage" {
+		return
+	}
+	ref, err := api.ParseDockerImageReference(from.Name)
+	if err != nil {
+		status.Status = invalidStatus("", fielderrors.NewFieldInvalid("from.name", from.Name, fmt.Sprintf("invalid name: %v", err)))
+		return
+	}
+	defaultRef := ref.DockerClientDefaults()
+	repoName := defaultRef.RepositoryName()
+	registryURL := defaultRef.RegistryURL()
+
+	key := repositoryKey{url: *registryURL, name: repoName}
+	repo := &importRepository{
+		Ref:         ref,
+		Registry:    &key.url,
+		Name:        key.name,
+		Insecure:    spec.ImportPolicy.Insecure,
+		MaximumTags: maximumTags,
+	}
+	importRepositoryFromDocker(ctx, repo, limiter)
+
+	if repo.Err != nil {
+		status.Status = imageImportStatus(repo.Err, "", "repository")
+		return
+	}
+
+	additional := []string{}
+	tagKey := manifestKey{repositoryKey: key}
+	for _, s := range repo.AdditionalTags {
```

Comment:
This seems straightforward, but I still think you should add a comment explaining that this only keeps the priority ordering correct  because the two lists are ordered and the `repo.Tags` are picked from the head of the list.